lib/grammers-client/src/client/net.rs

```diff
@@ -263,20 +284,85 @@ impl Client {
                         );
                         tokio::time::sleep(delay).await;
                         slept_flood = true;
-                        rx = self.0.request_tx.lock("invoke").enqueue(request);
+                        rx = enqueuer.lock("invoke").enqueue(request);
                         continue;
                     }
                     Err(e) => break Err(e),
                 },
                 Err(TryRecvError::Empty) => {
-                    self.step().await?;
+                    self.step_sender(sender, notifier).await?;
                 }
                 Err(TryRecvError::Closed) => {
                     panic!("request channel dropped before receiving a result")
                 }
             }
         }
     }
+    async fn connect_sender(&self, dc_id: i32) -> Result<Arc<FileDownloader>, InvocationError> {
+        let mut mutex = self.0.downloader_map.write().await;
+        debug!("Connecting new datacenter {}", dc_id);
+        return match connect_sender(dc_id, &self.0.config).await {
+            Ok((new_sender, new_tx)) => {
+                let new_downloader = Arc::new(FileDownloader {
+                    sender: AsyncMutex::new("dc_sender", new_sender),
+                    request_tx: Mutex::new("dc_enqueuer", new_tx),
+                    stepping_done: Notify::new(),
+                });
+
+                // export auth
+                let authorization = self.export_authorization(dc_id).await?;
+
+                // import into new sender
+                let request = tl::functions::auth::ImportAuthorization {
+                    id: authorization.id,
+                    bytes: authorization.bytes,
+                };
+                self.invoke_request(
+                    &request,
+                    &new_downloader.request_tx,
+                    &new_downloader.sender,
+                    &new_downloader.stepping_done,
+                )
+                .await?;
+
+                mutex.insert(dc_id, new_downloader.clone());
+                Ok(new_downloader.clone())
+            }
+            Err(e) => {
+                panic!("Cannot obtain new sender to datacenter {}", e)
```

Comment:
Do we really want a panic here, or should this return the error?

Reply:
No reason for panic 👍

Reply:
How should this ```AuthorizationError``` should be propagated ? it is not compatible with ```InvocationError```

Reply:
I am thinking in adding a new variant to ```InvocationError``` to signal for security issues

Reply:
The method should return `AuthorizationError` instead (because `InvocationError` is compatible with it, just not the other way around).

Reply:
My question was more like in terms of API breakage,  as this is method is part a chain of methods:
download_media (file.rs)-> next (file.rs) -> invoke_in_dc (net.rs) -> connect_sender (net.rs)
If I change it to return ```AuthorizationError``` the callers will also have to return it and so on

Reply:
marked as resolved, so keep panic?